types/context.go

```diff
@@ -111,8 +110,7 @@ func (c Context) BlockGasWanted() uint64                        { return c.block
 
 // clone the header before returning
 func (c Context) BlockHeader() cmtproto.Header {
-	msg := proto.Clone(&c.header).(*cmtproto.Header)
-	return *msg
+	return c.header
```

Comment:
this actually save significant memory allocations, because it's called many time.

Reply:
comment need update then, do we still need copy in [HeaderHash](https://github.com/crypto-org-chain/cosmos-sdk/blob/51b9e85da010920b501858434d12c7e9aa330003/types/context.go#L119)?

Reply:
it's still copied, just shallow copy.

Reply:
comment updated